cirq/ops/linear_combinations.py

```diff
@@ -351,6 +387,36 @@ def copy(self) -> 'PauliSum':
         factory = type(self)
         return factory(self._linear_dict.copy())
 
+    def matrix(self) -> np.ndarray:
+        """Reconstructs matrix of self from underlying Pauli operations.
+
+        Raises:
+            TypeError: if any of the gates in self does not provide a unitary.
+        """
+        num_qubits = len(self.qubits)
+        num_dim = 2**num_qubits
+        qubit_to_axis = {q: i for i, q in enumerate(self.qubits)}
+        result = np.zeros((2,) * (2 * num_qubits), dtype=np.complex128)
+        for vec, coeff in self._linear_dict.items():
+            op = _pauli_string_from_unit(vec)
+            identity = np.eye(num_dim,
+                              dtype=np.complex128).reshape(result.shape)
+            workspace = np.empty_like(identity)
+            axes = tuple(qubit_to_axis[q] for q in op.qubits)
+            u = protocols.apply_unitary(
+                op, protocols.ApplyUnitaryArgs(identity, workspace, axes))
+            result += coeff * u
```

Comment:
shouldn't the body of this loop be factored into a method on `PauliString`? Is the problem tensoring in the right number of identities to each PauliString? Presumably the `PauliString.matrix()` method should take a `qubits` argument that (a) sets the order of the qubits and (b) will tensor in appropriate identities if the string doesn't have explicit support on a given input qubit

Reply:
Good suggestion. Done.

Also, filed #3099 to make `matrix` more official and useful.